src/model.cpp

```diff
@@ -475,6 +475,45 @@ bool Model::hasImports() const
     return importsPresent;
 }
 
+void fixComponentUnits(const ModelPtr &model, const ComponentPtr &component)
+{
+    for (size_t v = 0; v < component->variableCount(); ++v) {
+        auto variable = component->variable(v);
+        if (variable->units() != nullptr) {
+            // Find the units in the model and switch out.
+            auto units = model->units(variable->units()->name());
+            if (units != nullptr) {
+                variable->setUnits(units);
+            }
+        }
+    }
+    for (size_t c = 0; c < component->componentCount(); ++c) {
+        fixComponentUnits(model, component->component(c));
+    }
+}
+
+void fixImportSourceUnits(const ImportSourcePtr &i1, ImportSourcePtr &i2)
+{
+    auto m2 = owningModel(i2);
+
+    // Go through all the imported units in this import source and update their sources.
+    for (size_t index = 0; index < i1->unitsCount(); ++index) {
+        auto u1 = i1->units(index);
+        auto u2 = m2->units(u1->name());
+        u2->setImportSource(i2);
+    }
+}
+
+void fixImportSourceComponents(const ImportSourcePtr &i1, ImportSourcePtr &i2)
+{
+    auto m2 = owningModel(i2);
+    for (size_t index = 0; index < i1->componentCount(); ++index) {
+        auto c1 = i1->component(index);
+        auto c2 = m2->component(c1->name(), true);
```

Comment:
See https://github.com/cellml/libcellml/pull/704/files#r482676716.

Reply:
I know we are dealing with a component here, and not a unit, but it's the same story: not sure we can only rely on the name of a component?

Reply:
What else would you use that's safer?

Reply:
Yeah, not sure to be honest. I was just thinking about the case where you have a (currently invalid) model with two components with the same name.

Reply:
I'm thinking that in this case we don't worry about invalid models. There seems to be no easy way to fix up these issues with cloning and having imported components or units with duplicate names is just going to screw up all efforts at trying to handle cloning a model correctly. Happy to leave that as a know feature and submit an issue to keep track of it. Assuming this one goes in like it is now.

Reply:
Ok, fair enough. I am happy with this PR (besides the case where two units/components have the same name).

Reply:
For completeness, #705 is the issue to track this for future developers to fix.